### PR TITLE
Add radix based policy resolver

### PR DESCRIPTION
**Pull request by nmvalera**
## PR description

Add implementation of a radix tree based privacy resolver.

It covers both
- aggregation procedure of a set of policies
- resolution procedue for a given operation 

## Changelog

- [ ] Add radix based Policy Resolver (https://github.com/consensys/quorum-key-manager/blob/master/CHANGELOG.md).

### CODE REVIEW

**Comment by nmvalera** [comment]
# Aggregation

Aggregation consists of storing statements with prefix patterns in radix structure so they can be resolved in a timely optimized manner when checking authorizations.

In particular, it avoids having to loop through all statements each time we check authorizations. 

We aggregate into
- one radix tree for resource path prefixes
- for each resource path prefix we have an action tree

Below is an example of aggregation for a 4 statements policy

Radix Policy Resolver

# Resolution

When resolving we visit only the parts of the tree that the operation is concerned about
- if meeting a single Deny the operation is "Denied"
- if not meeting at least once Allow the operation is "Not Allowed"

As soon as a "Deny" is met, we abort the visit of the tree and return a "Deny"

Below are some examples

# Base allowed operation

**Path**: path/bar/test 
**Action**: DoA-Test 

Radix Policy Resolver (1)

In this case

1. We start the visit at path prefix `*` which is a match for `path/bar/test `
2. We visit the corresponding action tree which has no match for `DoA-Test`
3. We continue paths visit up to `path/bar` which is a match
4. We visit the corresponding action tree up to `DoA*` which is a match  for `DoA-Test`
5. `DoA*` match is an `Allow`
6. No other match so we operation is allowed

File: src/auth/policy/resolver.go [hunk 1]
```diff
@@ -1,41 +1,349 @@
...
+	if res.allow == nil {
+		return fmt.Errorf("action %q on resource %q not allowed", res.op.Action, res.op.Path)
+	}
+
+	return nil
+}
+
+// actionsResolver is associated to every statement path
+// it aggregates all statement actions associated to the path into a radix tree
+type actionsResolver struct {
```
Comment (ggarri) [inline src/auth/policy/resolver.go:61]: Why couldn't you use `radix.Tree` to allocate also your `exactAction` all your effectResolver? As I can quickly see from go-radix each node can store a value where we would store the effectResolver, is that right?

File: src/auth/policy/resolver.go [hunk 2]
```diff
@@ -1,41 +1,349 @@
...
 	"github.com/consensys/quorum-key-manager/src/auth/types"
 )
 
+var (
+	wildCardSuffix = "*"
+)
+
 type Operation struct {
 	Action string
 	Path   string
```
Comment (ggarri) [inline src/auth/policy/resolver.go:17]: As far as I can see from our documentation an operation would involve two elements: an action and a resource for instance `key:Get` over `stores::stores/mystore/keys/*`. if so, should we rename `Path` by `ResourcePath` for clarity?

File: src/auth/policy/resolver.go [hunk 3]
```diff
@@ -1,41 +1,349 @@
...
+	allow *statement
+	deny  *statement
 }
 
-// Resolver is responsible to control whether an operation is authorized or not
-// depending on the set of policies attached to the resolver
-type Resolver struct {
-	// TODO: to be implemented
+func (res *result) Allowed() bool {
+	return res.deny == nil && res.allow != nil
```
Comment (ggarri) [inline src/auth/policy/resolver.go:44]: is it possible to obtain this resolt `res.deny == nil and res.allow == nil`? If so we will be denying access even if deny is not added. So that I believe we could simplify this method by:
```
unc (res *result) Allowed() bool {
	return res.allow != nil
```

File: src/auth/policy/resolver.go (hunk 1 above)
Comment (nmvalera) [reply src/auth/policy/resolver.go:61]: In the radix tree all keys are prefixes. If we also include the exact path into tree then `Sign` would also give access to `SignTransaction` and we don't want that.

File: src/auth/policy/resolver.go (hunk 2 above)
Comment (nmvalera) [reply src/auth/policy/resolver.go]: Updated.

File: src/auth/policy/resolver.go (hunk 3 above)
Comment (nmvalera) [reply src/auth/policy/resolver.go:44]: Yes this will happen when an operation has no `Deny` and no `Allow`. In which case the operation is not allowed (which is different than denied because there is a `Deny` statement)

File: src/auth/policy/resolver.go [hunk 4]
```diff
@@ -1,41 +1,349 @@
...
+	return r.isAuthorized(op)
+}
+
+func (r *RadixResolver) isAuthorized(op *Operation) (res *result) {
+	res = &result{
+		op: op,
+	}
+
+	// do we have an exact match for the path?
+	actionRslvr, hasExact := r.exactPath[op.Path]
```
Comment (ggarri) [inline src/auth/policy/resolver.go]: If I would have the following policies, and an user would try to do  `ACTION: keys:Get RESOURCE:stores::stores/mystore/keys/key1`, what would be the result?
```
- effect: Deny
        actions: 
            - keys:Get
        resource:
            - stores::stores/mystore/keys/*
- effect: Allow
        actions: 
            - keys:Get
            - keys:Sign
        resource:
            - stores::stores/mystore/keys/key1
```

Reading from left-to-right I would expect I would expect it to reject, reading right-to-left deny. I personally believe it should be left-to-right, Did you based you decision on how k8s does or any other standard?

File: src/auth/policy/resolver.go [hunk 5]
```diff
@@ -1,41 +1,349 @@
...
+	}
+}
+
+// RadixResolver allows to perform authorization checks
+
+// It is built from a set of Policy by aggregating all statements into some
+// radix trees structure allowing optimized authorization checks
+type RadixResolver struct {
+	// exactPath maps an exact path to an action resolver
+	exactPath map[string]*actionsResolver
```
Comment (ggarri) [inline src/auth/policy/resolver.go:210]: As far as I see we could we don't need `exactPath` and it can use radix.Tree to manage every scenario

File: src/auth/policy/resolver.go [hunk 6]
```diff
@@ -1,41 +1,349 @@
...
-func (res *Result) Allowed() bool {
-	return res.isAllowed
+type statement struct {
+	types.Statement
+	policy *types.Policy
 }
 
-func (res *Result) Error() error {
-	return res.err
+type result struct {
```
Comment (ggarri) [inline src/auth/policy/resolver.go:36]: Why did you need to store all this information as part of the result object. Isn't it enough to return a true or false as the result of resolving the access requested Operation ?

File: src/auth/policy/resolver.go [hunk 7]
```diff
@@ -1,41 +1,349 @@
...
 
-// Resolver is responsible to control whether an operation is authorized or not
-// depending on the set of policies attached to the resolver
-type Resolver struct {
-	// TODO: to be implemented
+func (res *result) Allowed() bool {
+	return res.deny == nil && res.allow != nil
+}
+
+func (res *result) Error() error {
```
Comment (ggarri) [inline src/auth/policy/resolver.go:47]: In which context did you intend to use this errors? This looks more like responsability of the Audit to define which kind of output message it should be added into the auditing logs but for the Store looking to allow or deny and operation this Error() does not seem useful.

File: src/auth/policy/resolver.go [hunk 8]
```diff
@@ -1,41 +1,349 @@
...
+	actionPrefix, _, hasPrefix := rslvr.prefixedActions.LongestPrefix(op.Action)
+	if !hasPrefix {
+		// no prefix so we return
+		return res
+	}
+
+	// we have a prefix
+	// we walk all matching prefix starting from root up to the longest matching prefix
+	// if we meet a single Deny we stop walking and return a Deny
+	rslvr.prefixedActions.WalkPath(actionPrefix, func(_ string, v interface{}) bool {
```
Comment (ggarri) [inline src/auth/policy/resolver.go:151]: Reading Radix docs (see below) this method appear to navigate top to bottom which implies you will be apply policies `left-to-right`, but using the exactPath you are applying the oppositive behavior of reading `right-to-left` (see comment here https://github.com/ConsenSys/quorum-key-manager/pull/216/files#r671375931) , is there something I am missing here? 

```
// WalkPath is used to walk the tree, but only visiting nodes
// from the root down to a given leaf. Where WalkPrefix walks
// all the entries *under* the given prefix, this walks the
// entries *above* the given prefix.
func (t *Tree) WalkPath(path string, fn WalkFn) {
```

File: src/auth/policy/resolver.go (hunk 1 above)
Comment (ggarri) [reply src/auth/policy/resolver.go:61]: Then it looks like Radix might not be a good fit since I would expect nodes to be split by a separator char, such as `/` or `:`. I would rather implement my own tree solution if that would be the case

As you describe you would have same issue with resources or action nesting other such as:
```
Action:
- Keys:List
- Keys:ListDeleted

Resources:
- stores::stores/mystoreOne/
- stores::stores/mystore
```

File: src/auth/policy/resolver.go (hunk 4 above)
Comment (nmvalera) [reply src/auth/policy/resolver.go]: In the example above it will "Deny"

File: src/auth/policy/resolver.go (hunk 1 above)
Comment (nmvalera) [reply src/auth/policy/resolver.go:61]: Not sure to understand.

Radix is used for prefixes (with wildcard last character  `*`) in the case above those are exact path.

File: src/auth/policy/resolver.go (hunk 5 above)
Comment (nmvalera) [reply src/auth/policy/resolver.go:210]: Hmmm...

First, I am not sure how we could handle cases where you have statements with both the exact path and prefix path (like `/store/keys*` and `/store/keys`).

Second, 
- storing in radix is more expensive in storage than a regular map (because we have intermediary nodes), we accept this extra cost to accelerate prefix pattern resolution
- for an exact path, a normal map is less expensive in storage and enables faster queries than a radix

So having both a radix for prefixes and a map for exact sounds the best option (btw this is also how it is implemented in Vault).

File: src/auth/policy/resolver.go (hunk 6 above)
Comment (nmvalera) [reply src/auth/policy/resolver.go:36]: To return extra information details about why an operation was rejected to a user. It helps users configure/debug their policies.

Typically when having a large number of policies, it would be almost impossible to properly configure/debug the policies.

Btw this is what AWS does.

**Comment by darioAnongba** [comment]
Hi! I started reviewing the Policy resolver. I'm probably not understanding this feature and I'm sorry to voice concerns now that it's already implemented but I already made this remark before. I consider this implementation to be completely over-engineered. Maybe I'm missing something but:
- Why are these policies needed in the first place ?
- What are the use cases where these are needed
- Why did we decide to go for such an implementation if (as far as I know) the different Vault solutions do not implement such fine-grained rules.

From my point of view this is not needed because the QKM only deals with creation and usage of keys and secrets:
- A secret can either be read or written. It makes no sense to allow someone to create a secret but not update it. Actions are only READ, WRITE.
- A key can either be used to sign or to encrypt. We either do the authorization per user with READ, CREATE, SIGN, ENCRYPT or on key creation (like AKV does) with SIGN, ENCRYPT

My point is that fine-grained authorization is not needed and will never be. It makes no sense to allow someone to SignTransaction but not SignEEATransaction. Now that we have a DB, we can store who is the user creating the resource and we can also store which operations are supported on a given resource, dropping this whole implementation entirely right?

cc: @ggarri  , @julien-marchand , @nmvalera

**Comment by darioAnongba** [comment]
Also setting policies via an external file sounds weird to me. At some point we will be able to create stores and nodes via API and we will store them in DB. Manifests will be deprecated. If so, setting of policies over created resouces should be done per resource or per user (with custom claims).

Again, I'm probably missing something here

File: src/auth/policy/resolver.go (hunk 1 above)
Comment (ggarri) [reply src/auth/policy/resolver.go:61]: But that is exactly what I am trying to avoid, using two independent flows for same feature. To mitigate this issue you describe above ` Sign would be treated as a prefix and would also give access to SignTransaction`, couldn't you just store as part of the node data (`effectResolver`) whether or not the path is a exactPath or a wildcarded one? so at the time you resolve you can deal with the issue of `Sign` vs `SignTransaction`

File: src/auth/policy/resolver.go (hunk 4 above)
Comment (ggarri) [reply src/auth/policy/resolver.go]: That is right, I would also expect deny due to your first statement, but in the method `isAuthorized(op)` in that particular operation you will fall into the `exactPath` branch of your code:
```
func (r *RadixResolver) isAuthorized(op *Operation) (res *result) {
	res = &result{
		op: op,
	}

	// do we have an exact match for the path?
	actionRslvr, hasExact := r.exactPath[op.Path]
	if hasExact {
		// if we have an exact match and it resolves to Deny we return Deny
		r := actionRslvr.isAuthorized(op)
		if r.deny != nil {
			return r
		}
		res = r
	}
```

what would make it `allowed` if I can read this code propertly

File: src/auth/policy/resolver.go (hunk 5 above)
Comment (ggarri) [reply src/auth/policy/resolver.go:210]: In terms of performance we are comparing a log(n) with O1, what IMO does not pay off the complexity of having two separate flows to maintain in the code and the additional complexity that means. I cannot see strong reasons to not use just radix for both alternatives, it might be worthy to have a chat about it to explore how it could be done.
